### PR TITLE
fix(python-plugin): update internal interpreter state immediately upon forking 

### LINKED ISSUE #2387: cpython: deadlocked workers if thread started before forking
## Problem

This issue has been observed with CPython 2.7 and above. If a thread is started by the Python interpreter in the master (parent) process before the master process forks a worker (child) process, a worker can fail to respawn when the child process does not gracefully shutdown. The reproduction below was used to investigate the failure to respawn and we found that child processes deadlock on acquiring the GIL after they are forked. 

This issue is a duplicate of:

- https://github.com/unbit/uwsgi/issues/1149
- https://github.com/unbit/uwsgi/issues/1369

## Reproduction

See source code for this reproduction at https://github.com/majorgreys/uwsgi-deadlock-workers/.

This issue can reliably be reproduced by starting a thread in a `sitecustomize.py` that enters a busy loop. This ensures that the current thread in the Python interpreter is not the main thread when uWSGI forks a child process. The result is an invalid thread state in the child process, specifically the GIL is locked and will not have been recreated.

- We install the debug version of CPython so that we can debug uWSGI and the Python plugin with gdb. 
- The reproduction consists of a simple hello world wsgi application and a `sitecustomize` that starts a thread which enters a busy loop for 5 seconds. 
- The `uwsgi.ini` enables a master process and threads along with one worker process with a max worker lifetime of 15 seconds. This allows us to more easily observe the failed attempt to respawn the worker process.

To build and run the reproduction with CPython 2.7 and uWSGI 2.0.20, run:

```
docker run --rm -it $(docker build -q .)
```

If the worker was respawning properly, we would expect the output of the process to be:

```
*** uWSGI is running in multiple interpreter mode ***
spawned uWSGI master process (pid: 35)
spawned uWSGI worker 1 (pid: 37, cores: 1)
spawned uWSGI http 1 (pid: 38)
[thread=140112973076224] run: finished.
worker 1 lifetime reached, it was running for 16 second(s)
Respawned uWSGI worker 1 (new pid: 39)
worker 1 lifetime reached, it was running for 16 second(s)
Respawned uWSGI worker 1 (new pid: 40)
```

However, uWSGI will hang waiting on the worker to respawn with uWSGI 2.0.2 in this reproduction:

```
*** uWSGI is running in multiple interpreter mode ***
spawned uWSGI master process (pid: 36)
spawned uWSGI worker 1 (pid: 38, cores: 1)
spawned uWSGI http 1 (pid: 39)
[thread=140082567735040] run: finished.
worker 1 lifetime reached, it was running for 16 second(s)
```

If we attach to the process with `gdb` we find that the cause is a deadlock on the GIL acquisition:

```
#0  futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x560bef494fd0) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
#1  do_futex_wait (sem=sem@entry=0x560bef494fd0, abstime=0x0) at sem_waitcommon.c:111
#2  0x00007f1348060988 in __new_sem_wait_slow (sem=sem@entry=0x560bef494fd0, abstime=0x0) at sem_waitcommon.c:181
#3  0x00007f13480609f9 in __new_sem_wait (sem=sem@entry=0x560bef494fd0) at sem_wait.c:42
#4  0x00007f13475f72c4 in PyThread_acquire_lock (lock=0x560bef494fd0, waitflag=1) at Python/thread_pthread.h:356
#5  0x0000560bee9d30d9 in gil_real_get ()
#6  0x0000560bee9cba6c in uwsgi_python_master_fixup ()
#7  0x0000560bee97f5c6 in uwsgi_respawn_worker ()
#8  0x0000560bee97d83b in master_loop ()
#9  0x0000560bee9bb78a in uwsgi_run ()
#10 0x0000560bee968e7e in main ()
```

## Root cause

In CPython, the thread state of a child process and its parent process must be updated after forking if there will be subsequent calls to the Python interpreter. This is handled by `PyOS_AfterFork_Child` in CPython 3+ and `PyOS_AfterFork` in CPython 2.7. 

In uWSGI, there is an attempt to acquire the GIL in the `master_fixup` Python plugin hook before `PyOS_AfterFork[_Child]` is called in `post_fork` hook. This leads to a deadlock which explains the failure for workers to respawn.

### PR DESCRIPTION
**Pull request by majorgreys**
Fixes #2387

This PR adds `pre_uwsgi_fork` and `post_uwsgi_fork` hooks to `uwsgi_plugin` in order to update the internal interpreter state of CPython immediately before and immediately after the call to `uwsgi_fork`. This follows the implementation of the standard library `os.fork()` in CPython. 

While these two hooks effectively replace existing code to update the internal interpreter state during the lifecycle of master and worker processes, we preserve the current behavior for uWSGI LTS. We instead provide a new option `py-call-uwsgi-fork-hooks` to enable the new feature.

We add test coverage for the issue of deadlocked workers raised in #2387 parameterized for relevant configurations options.

### CODE REVIEW

File: uwsgi.h [hunk 1]
```diff
@@ -1039,6 +1039,8 @@ struct uwsgi_plugin {
 	void *data;
 	void (*on_load) (void);
 	int (*init) (void);
+	void (*pre_uwsgi_fork) (void);
```
Comment (unbit) [inline uwsgi.h:1042]: Can you move it at the end of the structure for maintaining ABI?

File: core/uwsgi.c [hunk 2]
```diff
@@ -3338,14 +3338,6 @@ int uwsgi_start(void *v_argv) {
 		}
 	}
 
-	// master fixup
```
Comment (unbit) [inline core/uwsgi.c:3341]: you should leave this part in for not breaking other plugins

File: core/mule.c [hunk 3]
```diff
@@ -69,12 +69,6 @@ void uwsgi_mule(int id) {
 
 		uwsgi_close_all_sockets();
 
-		for (i = 0; i < 256; i++) {
```
Comment (unbit) [inline core/mule.c:72]: you should leave this part in for not breaking other plugins

**Comment by unbit** [comment]
@majorgreys thanks a lot, my only doubt (in addition to the review comments) is about the old (and ugly, obviously)  up.call_osafterfork. Basically it loses any meaning with your patch so i think we can safely remove this (at least in master). For LTS i would probably introduce this new behaviour (albeit definitely correct) as an option. I know it is annoying but LTS grants the user that we do not break/change old stuff (honestly i am not sure about the implications, very probably only app-managed signal management will be affected, and in a positive way). @xrmx any thought?

File: core/uwsgi.c (hunk 2 above)
Comment (majorgreys) [reply core/uwsgi.c:3341]: @unbit I did not see the `master_fixup` hook used elsewhere in the repository but that doesn't mean it's not used by a plugin a user might have, so I can add this back. Are you aware of such a plugin?

**Comment by majorgreys** [comment]
I've updated the change now to make it so the old behavior can be retained, specifically with regards to `call_osafterfork`. 

Given the needs of LTS, I'm inclined to start with the latter so it can be easily backported to the 2.0 branch and open a separate PR for the breaking changes so that they can end up on master. @unbit Does that work for what you had in mind for how to introduce these changes?

**Comment by majorgreys** [comment]
@unbit @xrmx I have updated the PR to introduce this as an opt-in feature and avoid any breaking changes so it can be backported into the LTS 2.0 branch.

**Comment by brettlangdon** [comment]
@unbit or @xrmx have you had a chance to look over this PR again? Is there anything else you would like us to update?

**Comment by majorgreys** [comment]
👋🏽 @unbit Any update on this PR? I addressed the concerns you raised with regards to merging this into LTS.

**Comment by xrmx** [comment]
Merged and broke master CI :)
Search for "ERROR: uWSGI did not start"
https://github.com/unbit/uwsgi/runs/5633992704?check_suite_focus=true

**Comment by majorgreys** [comment]
@xrmx Thanks for the heads up. I'll see if I can reproduce this locally.

**Comment by Superskyyy** [comment]
@majorgreys Hello, thanks for fixing the issue, but I was trying the new option `py-call-uwsgi-fork-hooks` and the following ignored exception prints out.
```

Exception ignored in: <function _after_at_fork_child_reinit_locks at 0x7fd01e7ecdc0>
Traceback (most recent call last):
  File "/home/gitpod/.pyenv/versions/3.8.13/lib/python3.8/logging/__init__.py", line 264, in _after_at_fork_child_reinit_locks
    _releaseLock()  # Acquired by os.register_at_fork(before=.
  File "/home/gitpod/.pyenv/versions/3.8.13/lib/python3.8/logging/__init__.py", line 232, in _releaseLock
    _lock.release()
RuntimeError: cannot release un-acquired lock
```

This was discussed here but remains unresolved, can you help to see if it is my env problem or something else? Really appreciated.

**Comment by majorgreys** [comment]
@Superskyyy I need to still review the linked issue but I can confirm that I can reproduce this error. I don't see it when `--py-call-uwsgi-fork-hooks` is not used along with `--enable-threads`.

#2179 also reported the same exception when `--py-call-osafterfork` was used.

Code for minimal reproduction: https://gist.github.com/majorgreys/3ee1716b107ad80226b12145ae87276a

**Pull request merged**